LandSandBoat/server
# Add missing mini-expansion key item trading logic to Squintrox Dryeyes

**_I affirm:_**

## What does this pull request do?

This is the partner to my Tenshodo Coffer PR. Squintrox logic is not fun to look at in this first commit, but I hope to utilize the setbit logic from tenshodo coffer and build a couple tables to shorten the core logic. 

It is tested and technically works, but I am not satisfied with the layout, just putting it here to get feedback on a path to the end result.

Edit: Testing is complete and I'm satisfied with the behavior

Notes about the logic for this NPC:
 - The order of page 1 `onTrigger` does NOT match the order of event update options and event finish options, so this is manually defined per option value. 
 - The order of page 2 `onTrigger` DOES match the order of event finish options, so this was simple to map `option - 299` to the `menuMetadata` index

## Steps to test these changes

This one is tricky to test as there's multiple layers of logic here:
- Trades:
  - Trading the required items for: `CRIMSON_KEY`, `WHITE_CORAL_KEY`, and `MOOGLE_KEY` are the simplest. 
  - Trade should only succeed if the `charVar` for that particular key item maps to a timestamp on the player's charvar less than now (this value is set to next jst midnight upon a successful trade)
- Page 1 key items:
  - Page 1 lists the key items available based on
    - The respective mini expansion being complete (mission log shows the `FIN` mission as active)
    - the corresponding timestamp in the player's `charVar` player variable.
    - player not currently having the key item
  - if the player chooses that key item, the requirements are checked:
    - if the player has all required key items (most require full list, `BIRD_KEY` only requires 3 of the 6 key items) they receive the key item and all req (and prereq) key items are removed
    - if the player doesn't meet the requirements, info is given:
      - for all but `BLACK_CORAL_KEY` and `MOOGLE_KEY`, a simple text is given stating the requirements for obtaining the repeat key item
        - `MOOGLE_KEY`: the particular enfeeblement kit required is chosen/listed
        - `BLACK_CORAL_KEY`: the particular zone for digging is chosen/listed
- Page 2 key items:
  - These are all prerequisite key items for completing the mini-expansion battles again to obtain the key item rewards
  - They are not displayed if any is true
    - The respective mini expansion is not complete (mission log shows the `FIN` mission as active)
    - you currently have _ALL_ the key items 
    - you're on cooldown due to the `charVar` timestamp on your player
  - If chosen, they cost a varied amount of gil/beastmen seals, with custom messages for each type of result

Various screenshots of testing

![image](https://github.com/LandSandBoat/server/assets/131182600/4d7f8420-0668-4aa4-8327-0fc69f1545e7)

![image](https://github.com/LandSandBoat/server/assets/131182600/b49f1a16-c54d-4e23-ae9c-ce4fed326d74)

![image](https://github.com/LandSandBoat/server/assets/131182600/118b0185-6cf2-4177-ba05-1d3fc4c9d57e)
